rosbag2/src/rosbag2/rosbag2.cpp

```diff
@@ -35,55 +41,139 @@ const char * ROS_PACKAGE_NAME = "rosbag2";
+      storage->create_topic(topic_name, topic_type);
+    }
+  }
+
+  if (subscriptions_.empty()) {
+    throw std::runtime_error("No topics could be subscribed. Abort");
+    return;
+  }
+
+  RCUTILS_LOG_INFO_NAMED(ROS_PACKAGE_NAME, "Waiting for messages...");
+  while (rclcpp::ok()) {
     rclcpp::spin(node);
   }
+  subscriptions_.clear();
+}
+
+std::map<std::string, std::string> Rosbag2::get_topics_with_types(
+  const std::vector<std::string> & topic_names, std::shared_ptr<rclcpp::Node> node)
+{
+  // TODO(Martin-Idel-SI): This is a short sleep to allow the node some time to discover the topic
+  // This should be replaced by an auto-discovery system in the future
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  auto topics = node->get_topic_names_and_types();
+
+  std::map<std::string, std::string> topic_names_and_types;
+  for (const auto & topic_name : topic_names) {
+    std::string complete_topic_name = topic_name;
+    if (topic_name[0] != '/') {
+      complete_topic_name = "/" + topic_name;
+    }
+    auto position = topics.find(complete_topic_name);
+    if (position != topics.end()) {
+      if (position->second.size() > 1) {
+        RCUTILS_LOG_ERROR_NAMED(
+          ROS_PACKAGE_NAME,
+          "Topic '%s' has several types associated. Only topics with one type are supported.",
+          position->first.c_str());
+      } else {
+        topic_names_and_types.insert({position->first, position->second[0]});
+      }
+    }
+  }
+  return topic_names_and_types;
+}
+
+std::shared_ptr<GenericSubscription>
+Rosbag2::create_subscription(
+  const std::function<void(std::string)> & after_write_action,
+  std::shared_ptr<rosbag2_storage::storage_interfaces::ReadWriteInterface> storage,
+  std::shared_ptr<Rosbag2Node> & node,
+  const std::string & topic_name, const std::string & topic_type) const
+{
+  auto subscription = node->create_generic_subscription(
+    topic_name,
+    topic_type,
+    [storage, topic_name, after_write_action](std::shared_ptr<rmw_serialized_message_t> message) {
+      auto bag_message = std::make_shared<rosbag2_storage::SerializedBagMessage>();
+      bag_message->serialized_data = message;
+      bag_message->topic_name = topic_name;
+      rcutils_time_point_value_t time_stamp;
+      int error = rcutils_system_time_now(&time_stamp);
+      if (error != RCUTILS_RET_OK) {
+        RCUTILS_LOG_ERROR_NAMED(
+          ROS_PACKAGE_NAME,
+          "Error getting current time. Error: %s", rcutils_get_error_string_safe());
+      }
+      bag_message->time_stamp = time_stamp;
+
+      storage->write(bag_message);
+      if (after_write_action) {
+        after_write_action(topic_name);
+      }
+    });
+  return subscription;
 }
 
-void Rosbag2::play(const std::string & file_name, const std::string & topic_name)
+void Rosbag2::play(const std::string & file_name)
 {
   rosbag2_storage::StorageFactory factory;
   auto storage = factory.open_read_only(file_name, "sqlite3");
 
   if (storage) {
```

Comment:
Nothing :grin:. The storage factory already complains if the storage cannot be loaded.

If we want to have more elaborate error handling I would postpone that until the `rosbag2_transport` has been introduced as we would have a CLI integration then. Then it would be easier to think about feedback to the user.